include/RAJA/exec-cuda/scan_cuda.hxx

```diff
@@ -72,73 +71,125 @@ namespace detail
 namespace scan
 {
 
+inline cub::CachingDeviceAllocator& getAllocator() {
+  static cub::CachingDeviceAllocator allocator(true);
+  return allocator;
+}
+
 /*!
         \brief explicit inclusive inplace scan given range, function, and
    initial value
 */
-template <typename InputIter, typename Function, typename T>
-void inclusive_inplace(const ::RAJA::cuda_exec_base&,
+template <size_t BLOCK_SIZE, bool Async,
+          typename InputIter, typename Function, typename T>
+void inclusive_inplace(const ::RAJA::cuda_exec<BLOCK_SIZE, Async>&,
                        InputIter begin,
                        InputIter end,
                        Function binary_op,
                        T init)
 {
-  ::thrust::inclusive_scan(::thrust::device, begin, end, begin, binary_op);
-  cudaDeviceSynchronize();
+  int num_items = std::distance(begin, end);
+  // Determine temporary device storage requirements
+  void     *d_temp_storage = NULL;
+  size_t   temp_storage_bytes = 0;
+  cudaErrchk(cub::DeviceScan::InclusiveScan(d_temp_storage, temp_storage_bytes, begin, begin, binary_op, num_items));
+  // Allocate temporary storage
+  cudaErrchk(getAllocator().DeviceAllocate(&d_temp_storage, temp_storage_bytes));
+  // Run
+  cudaErrchk(cub::DeviceScan::InclusiveScan(d_temp_storage, temp_storage_bytes, begin, begin, binary_op, num_items));
+  // Free temporary storage
+  cudaErrchk(getAllocator().DeviceFree(d_temp_storage));
+
+  RAJA_CUDA_CHECK_AND_SYNC(Async);
 }
 
 /*!
         \brief explicit exclusive inplace scan given range, function, and
    initial value
 */
-template <typename InputIter, typename Function, typename T>
-void exclusive_inplace(const ::RAJA::cuda_exec_base&,
+template <size_t BLOCK_SIZE, bool Async,
+          typename InputIter, typename Function, typename T>
+void exclusive_inplace(const ::RAJA::cuda_exec<BLOCK_SIZE, Async>&,
                        InputIter begin,
                        InputIter end,
                        Function binary_op,
                        T init)
 {
-  ::thrust::exclusive_scan(
-      ::thrust::device, begin, end, begin, init, binary_op);
-  cudaDeviceSynchronize();
+  int num_items = std::distance(begin, end);
+  // Determine temporary device storage requirements
+  void     *d_temp_storage = NULL;
+  size_t   temp_storage_bytes = 0;
+  cudaErrchk(cub::DeviceScan::ExclusiveScan(d_temp_storage, temp_storage_bytes, begin, begin, binary_op, init, num_items));
+  // Allocate temporary storage
+  cudaErrchk(getAllocator().DeviceAllocate(&d_temp_storage, temp_storage_bytes));
+  // Run
+  cudaErrchk(cub::DeviceScan::ExclusiveScan(d_temp_storage, temp_storage_bytes, begin, begin, binary_op, init, num_items));
```

Comment:
Oh that is weird. I didn't realize the `// Determine temporary ...` comment applied to those three lines.